File: quickstart_guides/linked_lists/java/PairsWithSum.java
/*
 * Title: Pairs with sum
 *
 * Problem:
 *  Given an array of integers, and a number ‘sum’, find the number of pairs of
 *  integers in the array whose sum is equal to ‘sum’.
 *
 * Execution: javac PairsWithSum.java && java PairsWithSum
 */
public class PairsWithSum {

    public static int pairsWithSum(int[] arr, int sum) {
        int count = 0;

        // Consider all possible pairs and check their sums
        for (int i = 0; i < arr.length; i++)
            for (int j = i + 1; j < arr.length; j++)
                if ((arr[i] + arr[j]) == sum)
                    count++;

        return count;
    }
    public static void main(String args[]) {
        int[] arr = { 1, 5, 7, -1, 5 };
        int sum = 6;
        assert pairsWithSum(arr, sum) == 3;
    }
}

File: quickstart_guides/linked_lists/python/pairs_with_sum.py
"""
Title: Pairs with sum.

Problem:
    Given an array of integers, and a number ‘sum’, find the number of pairs of
    integers in the array whose sum is equal to ‘sum’.

Execution: python pairs_with_sum.py
"""
from typing import List
import unittest


def pairs_with_sum(arr: List[int], s: int) -> int:
    """
    Returns number of pairs in arr[0..n-1] with sum equal to 'sum'
    """
    # Initialize result.
    count = 0
    # Consider all possible pairs and check their sums.
    for i in range(len(arr)):
        for j in range(i + 1, len(arr)):
            if arr[i] + arr[j] == s:
                count += 1
    return count


class TestPairsWithSum(unittest.TestCase):
    """Unit test for pairs_with_sum."""

    def test_1(self):
        x = [1, 5, 7, -1, 5]
        self.assertEqual(pairs_with_sum(x, 6), 3)


if __name__ == "__main__":
    unittest.main()
